backend/protondrive/protondrive.go

```diff
@@ -0,0 +1,1040 @@
+//
+// This should be implemented by the backend and will be called by the
+// dircache package when appropriate.
+//
+// CreateDir makes a directory with pathID as parent and name leaf
+func (f *Fs) CreateDir(ctx context.Context, pathID, leaf string) (string, error) {
+	/* f.opt.Enc.FromStandardName(leaf) not required since the DirCache only process sanitized path */
+
+	var newID string
+	var err error
+	if err = f.pacer.Call(func() (bool, error) {
+		newID, err = f.protonDrive.CreateNewFolderByID(ctx, pathID, leaf)
+		return shouldRetry(ctx, err)
+	}); err != nil {
+		return "", err
+	}
+
+	return newID, err
+}
+
+// Put in to the remote path with the modTime given of the given size
+//
+// When called from outside an Fs by rclone, src.Size() will always be >= 0.
+// But for unknown-sized objects (indicated by src.Size() == -1), Put should either
+// return an error or upload it properly (rather than e.g. calling panic).
+//
+// May create the object even if it returns an error - if so
+// will return the object and the error, otherwise will return
+// nil and the error
+func (f *Fs) Put(ctx context.Context, in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) (fs.Object, error) {
+	size := src.Size()
+	if size < 0 {
+		return nil, ErrCanNotUploadFileWithUnknownSize
+	}
+
+	existingObj, err := f.NewObject(ctx, src.Remote())
+	switch err {
+	case nil:
+		// object is found
+		return existingObj, existingObj.Update(ctx, in, src, options...)
+	case fs.ErrorObjectNotFound:
+		// object not found, so we need to create it
+		return f.PutUnchecked(ctx, in, src)
+	default:
+		// real error caught
+		return nil, err
+	}
+}
+
+// Creates from the parameters passed in a half finished Object which
+// must have setMetaData called on it
+//
+// Returns the object, leaf, directoryID and error.
+//
+// Used to create new objects
+func (f *Fs) createObject(ctx context.Context, remote string, modTime time.Time, size int64) (*Object, error) {
+	//                 ˇ-------ˇ filename
+	// e.g. /root/a/b/c/test.txt
+	//      ^~~~~~~~~~~^ dirPath
+
+	// Create the directory for the object if it doesn't exist
+	_, _, err := f.dirCache.FindPath(ctx, f.sanitizePath(remote), true)
+	if err != nil {
+		return nil, err
+	}
+
+	// Temporary Object under construction
+	obj := &Object{
+		fs:           f,
+		remote:       remote,
+		size:         size,
+		originalSize: nil,
+		id:           "",
+		modTime:      modTime,
+		mimetype:     "",
+		link:         nil,
+	}
+	return obj, nil
+}
+
+// Put in to the remote path with the modTime given of the given size
+//
+// May create the object even if it returns an error - if so
+// will return the object and the error, otherwise will return
+// nil and the error
+//
+// May create duplicates or return errors if src already
+// exists.
+func (f *Fs) PutUnchecked(ctx context.Context, in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) (fs.Object, error) {
```

Comment:
If proton can't have duplicates you don't have to implement this. 

I'm not sure whether it can or not or maybe only sometimes!

Reply:
Proton's backend doesn't allow duplicated file names. So I will have to remove this function entirely I believe, right?

Sorry for missing the case, as I was following other backends and didn't realize this is specific to this "duplicated filename case"!

Reply:
Removed as requested!